plugins/tactics/src/Ide/Plugin/Tactic.hs

```diff
@@ -71,15 +81,66 @@ descriptor plId = (defaultPluginDescriptor plId)
               (tacticCmd $ commandTactic tc))
               [minBound .. maxBound]
     , pluginCodeActionProvider = Just codeActionProvider
+    , pluginRules = tacticRules
     }
 
+data GetMetaprogram = GetMetaprogram
+  deriving stock (Show, Generic, Eq, Ord)
+  deriving anyclass (Hashable, NFData, Binary)
+
+type instance RuleResult GetMetaprogram = Metaprogram
+
+data GetMetaprogramCache = GetMetaprogramCache
+  deriving stock (Show, Generic, Eq, Ord)
+  deriving anyclass (Hashable, NFData, Binary)
+
+type instance RuleResult GetMetaprogramCache = MetaprogramCache
+
+tacticRules :: Rules ()
+tacticRules = do
+  defineEarlyCutoff $ \GetMetaprogram nfp -> do
+    mtime <- use GetModificationTime nfp
+    contents <- liftIO $ T.readFile $ fromNormalizedFilePath nfp
+    pure
+      $ (Just $ BS.pack $ show mtime ,)
+      $ case parseMetaprogram (fromNormalizedFilePath nfp) contents of
+          Left err ->
+            ( [ (nfp
+                , ShowDiag
+                , Diagnostic
+                    (Range (Position 0 0)
+                           (Position 0 0))
+                    Nothing
+                    Nothing
+                    Nothing
+                    (T.pack err)
+                    Nothing
+                    Nothing
+                )
+              ]
+            , Nothing
+            )
+          Right mp -> ([], Just mp)
+  define $ \GetMetaprogramCache _ -> do
+    files <- liftIO getKnownFiles
```

Comment:
`GetMetaprogramCache` won't be recomputed when the result of `getKnownFiles` changes.

Also, there's already a function called `getKnownFiles` in ghcide, you should rename it to avoid confusion.

Reply:
Good catch. What's the right way to discover new files in shake?

Reply:
Create an early cutoff rule for `getKnownFiles`, and use `alwaysRerun`.

Reply:
Although that might not be such a good idea, since it would call `ls` every time a codeaction is generated, right?

Reply:
I guess you need to register a file watcher with the client, like we were discussing earlier on IRC.